Davilu/component order

davilu opened the pull request:
Added logic to send notification to MQTT broker while EHC `AuthAgentProtocolHead` boot up;
Abstracted `AbstractNotificationHandler` for any component to send notification to MQTT broker

vipeller commented on edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/authentication/AuthAgentProtocolHead.cs:
```diff
@@ -121,6 +132,33 @@ public void Dispose()
+            IEnumerable<Message> messages = new[] { new Message(Topic, Payload) };
+            return Task.FromResult(messages);
+        }
+
```
I think it is too much logic here in the protocol head class just to be able to handle both situations when the connector is connected or not connected. I feel that this should not be the concern of this class. Can it be implemented the way that this (and the ScopeIdentities) would not care about the connected state - they just call a 'SendMessageWheneverYouCanAsync()' method with the baked message (topic and binary content, just like connector.SendAsync(), and then the Connector would contain the logic that once it is connected, it sends the messages out? In that case we don't need these three methods here and in the ScopeIdentities, and both AuthAgentProtocolHead and ScopeIdentities would have a very narrow code to implement the notification. Also, we could get rid of the base class. 
The connector class could store these messages in a MqttPublishInfo it already uses (so we could get rid of the Message class introduced here), and once it gets connected, it fires out the collected messages.

davilu replied:
The 3 API is very clear: prepare what message need to be send. 

- `ConvertStoredNotificationsToMessagesAsync` is based on if `sendNotificationWhenConnected`, return the notification message.
- `ConvertNotificationToMessagesAsync`is return the notification message.
- `StoreNotificationAsync` is ask to store state while notification message can't be sent(connection to broker is not established).

davilu commented on edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/authentication/AuthAgentProtocolHead.cs:
```diff
@@ -2,27 +2,34 @@
         readonly AuthAgentProtocolHeadConfig config;
         readonly object guard = new object();
 
+        bool _started;
```
I'm not quite understand your point. AFAIK, `StopAsync` will be called only when we bring down EHC. But if EHC is down, connection to MQTT broker is down. Even in future we allow restart, since `_started` is `false`, no message will be send until `StartAsync` was called again. Only gap is since `IMqttBrokerConnector` has no API to tell disconnection, if it connects and disconnect and after `StartAsync`, it will still try to send message through `IMqttBrokerConnector`.

davilu replied:
Modified to always send out notification while connected if `AuthAgentProtocolHead` is started

varunpuranik commented:
@davilu - nit: can you pls add description on the changes included in this PR? 
Also, pls make sure to target it to master instead of iiot.

davilu commented:
> @davilu - nit: can you pls add description on the changes included in this PR?
> Also, pls make sure to target it to master instead of iiot.

Modified target branch and added descriptions.

varunpuranik commented on edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/NotificationHandler.cs:
```diff
@@ -0,0 +1,162 @@
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+
+    public class NotificationHandler<T>
```
What is this for?

davilu replied:
A tool class for any notification need to be sent to mqtt broker

davilu commented:
/azp run

richma-ms commented:
/azp run

richma-ms commented on edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceScopeIdentitiesCache.cs:
```diff
@@ -342,6 +358,11 @@ async Task<Option<ServiceIdentity>> GetServiceIdentityInternal(string id)
+            if (this.lifeCycleSignal.IsCancellationRequested)
+            {
+                return;
+            }
```
seems like this should be consistent with HandleNewServiceIdentity, either we check cancellation in both, or neither.

davilu replied:
Good catch. Missed `HandleNewServiceIdentity`

richma-ms commented on edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceScopeIdentitiesCache.cs:
```diff
@@ -206,9 +213,11 @@ public void Dispose()
             this.encryptedStore?.Dispose();
             this.refreshCacheTimer?.Dispose();
-            this.refreshCacheTask?.Dispose();
+            this.lifeCycleSignal.Cancel();
```
>this.lifeCycleSignal.Cancel(); [](start = 12, length = 30)

hmm...this just signals cancel but doesn't wait.  What happens when background task gets disposed now?

davilu replied:
The background task will finish by itself. It could wait `refreshCacheTask` finish here if would like to guarantee the task quit. But block `Dispose` is not a ggod idea.

richma-ms commented on edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs:
```diff
@@ -20,97 +13,41 @@ public class ScopeIdentitiesHandler : IMessageProducer
-                deviceScopeIdentitiesCache.ServiceIdentitiesUpdated +=
-                    async (sender, serviceIdentities) => await this.ServiceIdentitiesUpdated(serviceIdentities);
+            var deviceScopeIdentitiesCache = await this.deviceScopeIdentitiesCacheGetter;
+            deviceScopeIdentitiesCache.ServiceIdentitiesUpdated += async (sender, serviceIdentities) => await this.notificationHandler.NotifyAsync(serviceIdentities);
```
>ServiceIdentitiesUpdated [](start = 39, length = 24)

ServiceIdentityRemoved should also be hooked up (I think this was missing before as well)

richma-ms commented:
>             await Task.Delay(TimeSpan.FromSeconds(refreshDelaySec));

Can you add the +1 sec timer fix here as well? (it's basically the same test logic)

---
Refers to: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.Core.Test/DeviceScopeIdentitiesCacheTest.cs:804 in 10fd0bd. [](commit_id = 10fd0bd006c2537fa9070856a5186a28a2e65ab5, deletion_comment = False)

richma-ms commented on edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/authentication/AuthAgentProtocolHead.cs:
```diff
@@ -133,7 +139,10 @@ enum EventIds
+                StartedWhenAlreadyRunning,
+                NotifyingBroker,
+                NotifiedBroker,
+                NotifyBrokerFailed
```
these don't seem used anymore

davilu replied:
Will remove them.

richma-ms commented on edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/authentication/AuthAgentProtocolHead.cs:
```diff
@@ -66,15 +68,15 @@ public async Task StartAsync()
             Events.Closing();
-
             Option<IWebHost> hostToStop;
+            this.startSignal.Reset();
```
>this.startSignal.Reset(); [](start = 12, length = 25)

this should probably come after the call to StopAsync, otherwise there's a window of inconsistency.  Not an issue right now but could lead to bugs in the future.

davilu replied:
Will do. Nice catch.

dylanbronson commented on edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/NotificationHandler.cs:
```diff
@@ -0,0 +1,152 @@
+        }
+    }
+
+    public class Message
```
Should we rename this BrokerMessage? Message is already overloaded in EdgeHub

davilu replied:
done
